speaker/bgp_controller.go

```diff
@@ -366,7 +382,40 @@ func (c *bgpController) updateAds() error {
 		if err := peer.session.Set(ads...); err != nil {
 			return err
 		}
+		for _, ad := range ads {
+			adSvcs := pfxToSvc[ad.Prefix.String()]
+			for svc := range adSvcs {
+				if _, ok := newActiveAds[svc]; !ok {
+					newActiveAds[svc] = sets.New[string](peer.cfg.Name)
+					continue
+				}
+				newActiveAds[svc].Insert(peer.cfg.Name)
+			}
+		}
 	}
+
+	changedSvcs := []string{} // the services that their advs changed
+	for svc, oldPeers := range oldActiveAds {
+		newPeers, ok := newActiveAds[svc]
+		if !ok {
+			changedSvcs = append(changedSvcs, svc)
+			continue
+		}
+		if !oldPeers.Equal(newPeers) {
+			changedSvcs = append(changedSvcs, svc)
+		}
+	}
+	for svc := range newActiveAds {
+		_, ok := oldActiveAds[svc]
+		if !ok {
+			changedSvcs = append(changedSvcs, svc)
+			continue
+		}
+	}
+	for _, k := range changedSvcs {
+		c.adsChangedCallback(k)
```

Comment:
Is there a unit-test for that? I mean If I comment that line, will any test fail?

Reply:
added